src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -2690,6 +2690,34 @@ LogicalResult ONNXCastOp::inferShapes(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// CastLike
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXCastLikeOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  ShapedType inputType = input().getType().dyn_cast<RankedTensorType>();
+  if (!inputType) {
```

Comment:
Any reason that the traditional 
```c++
if (!hasShapeAndRank(input())) {
     return success();
   }
```
is not used here?

Do you want to propagate the `unranked` type? Do you have a specific need for the unranked? My understanding is that onnx-mlir pretty much can't do anything with unranked at this time. But that might be only in our use case (which is lowering to actual code).

Reply:
@chentong319 I see that some shape inference simply propagate the input types to the output, but some other just report `success` when rank and shape are not yet avail. Is that ok?

@philass this question is not about your code, just that it made me realize that we are not entirely consistent. If we wanted to change how we do things, it would come to a different PR, so you are fine.

Reply:
Some really old code did not consider the case of unranked tensor. If the input is unranked, shape inference can do nothing other than return. There is no check when the input type is copied to the output type. The copy may overwrite the shape info coming with the model (such as partial static info).

Reply:
I'd be happy to look into improving some of the Shape Inference for better consistency in other PR